implemented security

JohnGrey17 opened the pull request:
- spring-boot-starter-security
- SecurityConfig
- PasswordEncoder
- SecurityFilterChain
- UserDetails
- UserDetailsService
- RoleRepository
- Set<Role> roles
- @PreAuthorize Created changLog
- added roles table
- added user_roles table
- insert users into the table
- assign roles for users

JohnGrey17 committed:
added
- spring-boot-starter-security
- SecurityConfig
- PasswordEncoder
- SecurityFilterChain
- UserDetails
- UserDetailsService
- RoleRepository
- Set<Role> roles
- @PreAuthorize
Created changLog
- added roles table
- added user_roles table
- insert users into the table
- assign roles for users

Nikname2303 commented on src/main/java/org/example/bookstore/config/SecurityConfig.java:
```diff
@@ -0,0 +1,43 @@
+                .csrf(AbstractHttpConfigurer::disable)
+                .authorizeHttpRequests(
+                        auth -> auth
+                                .requestMatchers("/auth/**", "/error","/swagger-ui/**")
```
make each from new line

Nikname2303 commented on src/main/java/org/example/bookstore/controller/AuthenticationController.java:
```diff
@@ -23,4 +23,5 @@ public UserResponseDto registerUser(@RequestBody @Valid UserRegistrationRequestD
             throws RegistrationException {
         return userService.register(requestDto);
     }
+
```
-

Nikname2303 commented on src/main/java/org/example/bookstore/controller/BookControllerImpl.java:
```diff
@@ -31,42 +32,48 @@ public class BookControllerImpl {
     }
 
     @GetMapping("/{id}")
+    @PreAuthorize("hasRole('ROLE_USER')")
```
Why only user cane see that? make access for user and admin

Nikname2303 commented on src/main/java/org/example/bookstore/controller/HelloController.java:
```diff
@@ -0,0 +1,12 @@
+package org.example.bookstore.controller;
+
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class HelloController {
+    @GetMapping("/hello")
+    public String sayHello() {
+        return "hello";
+    }
+}
```
I'm not sure we really need this

Nikname2303 commented on src/main/java/org/example/bookstore/repository/user/UserRepository.java:
```diff
@@ -1,10 +1,14 @@
 public interface UserRepository extends JpaRepository<User,Long> {
     boolean existsByEmail(String email);
+
+    Optional<User> findByEmail(@Param("email") String email);
```
you can siplify that, just write Optional<User> findByEmail(String email) - its the same thing

Nikname2303 commented on src/main/java/org/example/bookstore/security/CustomUserDetailService.java:
```diff
@@ -0,0 +1,21 @@
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+        return userRepository.findByEmail(username).orElseThrow(
+                () -> new UsernameNotFoundException("Can`t find user by username " + username));
```
rename variable for search on email for better understanding

Nikname2303 reviewed (changes requested):
Good job, just a few minor comments

JohnGrey17 committed:
Fixed solution

andrii-hoienko commented on src/main/java/org/example/bookstore/controller/BookControllerImpl.java:
```diff
@@ -31,42 +32,48 @@ public class BookControllerImpl {
     private final BookService bookService;
 
     @GetMapping
+    @PreAuthorize("hasRole('ROLE_USER')")
```
Why admin can get book by id, but cant find all books?

andrii-hoienko commented on src/main/java/org/example/bookstore/controller/BookControllerImpl.java:
```diff
@@ -31,42 +32,48 @@ public class BookControllerImpl {
     }
 
+    @GetMapping("/search")
+    @PreAuthorize("hasRole('ROLE_USER')")
```
Why admin can get book by id, but cant find search books?

andrii-hoienko commented on src/main/java/org/example/bookstore/service/impl/user/UserServiceImpl.java:
```diff
@@ -24,7 +26,7 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
         User user = userMapper.toModel(requestDto);
-        User savedUser = userRepository.save(user);
-        return userMapper.toDto(savedUser);
+        user.setPassword(passwordEncoder.encode(user.getPassword()));
+        return userMapper.toDto(userRepository.save(user));
```
Is it ok that use will not have a role after registration?

andrii-hoienko reviewed (changes requested):
Nice, left some questions and comments

JohnGrey17 committed:
fixed solution
*add default roll (USER) for all new user in service layer
*Updated Search,getBookById,GetAllBooks methods. added admin role

JohnGrey17 committed:
fixed solution
*add default roll (USER) for all new user in service layer
*Updated Search,getBookById,GetAllBooks methods. added admin role

karma-o commented on src/main/java/org/example/bookstore/config/SecurityConfig.java:
```diff
@@ -0,0 +1,45 @@
+                        auth -> auth
+                                .requestMatchers("/auth/**",
+                                        "/error",
+                                        "/swagger-ui/**")
```
I think you also need to add something like `/v3/api-docs/**` in order for swagger to work

karma-o commented on src/main/java/org/example/bookstore/controller/BookControllerImpl.java:
```diff
@@ -31,42 +32,48 @@ public class BookControllerImpl {
 
     @PostMapping
     @ResponseStatus(HttpStatus.CREATED)
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
```
we need to add `ROLE_` when we use `hasAuthorithy()` but here we already specified that we are checking exactly role.
Same for other similar cases
```suggestion
    @PreAuthorize("hasRole('ADMIN')")
```

karma-o commented on src/main/java/org/example/bookstore/repository/role/RoleRepository.java:
```diff
@@ -0,0 +1,8 @@
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface RoleRepository extends JpaRepository<Role, Long> {
+    Role getRoleById(Long id);
```
* we can not really guarantee that we will always have a result
* I prefer to use `find` instead of get wich implies that we are **trying** to get it but d not guarantee the result, when **get** implies that we are certain that we will get what we are looking for 
* how will you know the id to fetch the role by? Hardcode is not the best scenario because we can have different id generation strategies and types, but role name will always be constant
```suggestion
    Optional<Role> findRoleByName(RoleName name);
```

karma-o commented on src/main/java/org/example/bookstore/service/impl/user/UserServiceImpl.java:
```diff
@@ -24,7 +30,9 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
-        User savedUser = userRepository.save(user);
-        return userMapper.toDto(savedUser);
+        user.setPassword(passwordEncoder.encode(user.getPassword()));
+        Role roleById = roleRepository.getRoleById(1L);
```
```suggestion
        Role userRole = roleRepository.findRoleByName(RoleName.USER);
```

karma-o commented on src/main/resources/db/changelog/changes/07-add-roles-for-users.yaml:
```diff
@@ -0,0 +1,41 @@
+            columns:
+              - column:
+                  name: user_id
+                  value: "1"
+              - column:
+                  name: role_id
+                  value: "1"
```
I'd recommend using subselects to get by email instead of hardcoding ids here and in other similar places

karma-o reviewed (approved):
Looks good enough! But please take a look at my comments before merging

JohnGrey17 committed:
Fixed
*Role definition in Book controller
*method findRoleByName instead getRoleById (repository,service layer)
